pkg/storage/streams/store.go

```diff
@@ -269,12 +364,54 @@ func (lr *lazySegmentRanger) Size() int64 {
 
 // Range implements Ranger.Range to be lazily connected
 func (lr *lazySegmentRanger) Range(ctx context.Context, offset, length int64) (io.ReadCloser, error) {
+	cipher := eestream.Cipher(lr.encType)
+
 	if lr.ranger == nil {
-		rr, _, err := lr.segments.Get(ctx, lr.path)
+		rr, m, err := lr.segments.Get(ctx, lr.path)
+		if err != nil {
+			return nil, err
+		}
+		encryptedEncKey := m.Data
+		e, err := cipher.Decrypt(encryptedEncKey, lr.derivedKey, lr.startingNonce)
+		if err != nil {
+			return nil, err
+		}
+		var encKey eestream.GenericKey
+		copy(encKey[:], e)
+		decrypter, err := cipher.NewDecrypter(&encKey, lr.startingNonce, lr.encBlockSize)
 		if err != nil {
 			return nil, err
 		}
-		lr.ranger = rr
+
+		var rd ranger.Ranger
+		if rr.Size()%int64(decrypter.InBlockSize()) != 0 {
+			reader, err := rr.Range(ctx, 0, rr.Size())
+			if err != nil {
+				return nil, err
+			}
+			cipherData, err := ioutil.ReadAll(reader)
+			if err != nil {
+				return nil, err
+			}
+			data, err := cipher.Decrypt(cipherData, &encKey, lr.startingNonce)
+			if err != nil {
+				return nil, err
+			}
+			rd = ranger.ByteRanger(data)
+			lr.ranger = rd
+		} else {
+			rd, err = eestream.Transform(rr, decrypter)
+			if err != nil {
+				return nil, err
+			}
+
+			paddedSize := rd.Size()
+			rc, err := eestream.Unpad(rd, int(paddedSize-lr.Size())) // int64 -> int; is this a problem?
```

Comment:
Can we remove this comment too?

The difference between the padded size and the actual size of the segment should be a really small number (less than an encryption block size). So it shouldn't be a problem to fit into the `int` type.